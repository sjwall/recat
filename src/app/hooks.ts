import { useMemo } from 'react'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { selectCurrentUser } from '../features/auth/authSlice'
import type { RootState, AppDispatch } from './store';

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export const useAuth = () => {
  const user = useSelector(selectCurrentUser)

  return useMemo(() => ({ user }), [user])
}
