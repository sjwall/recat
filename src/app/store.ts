import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import authReducer from '../features/auth/authSlice';
import { authApi } from '../features/auth/authAPI';
import imageReducer from '../features/thecatapi/thecatapiSlice';
import { imageApi } from '../features/thecatapi/thecatapiAPI';

export const store = configureStore({
  reducer: {
    auth: authReducer,
    images: imageReducer,
    [authApi.reducerPath]: authApi.reducer,
    [imageApi.reducerPath]: imageApi.reducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
