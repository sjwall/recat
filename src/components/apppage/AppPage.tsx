import React, { ReactNode } from 'react'
import { Link } from "react-router-dom";
import { useTranslation } from "react-i18next";
import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import FileUpload from '@mui/icons-material/FileUpload';
import ExitToApp from '@mui/icons-material/ExitToApp';
import { User } from '../../features/auth/user.model';
import { useAppDispatch } from '../../app/hooks';
import { logout } from '../../features/auth/authSlice';

/**
 * Properties for AppPage.
 */
export type AppPageProps = {
  /**
   * Children to display below app bar.
   */
  children: ReactNode

  /**
   * Title to display in app bar.
   */
  title: string

  /**
   * User to display in app bar.
   */
  user: User
}

/**
 * Template page.
 *
 * @param props Properties.
 * @returns React component.
 */
export function AppPage(props: AppPageProps) {
  const { t } = useTranslation();
  const { children, title, user } = props;
  const dispatch = useAppDispatch();
  const handleLogout = () => {
    dispatch(logout())
  };

  return <>
    <Box sx={{ display: 'flex', flexDirection: 'column' }}>
      <AppBar position="static">
        <Toolbar>
          <Link to="/upload">
            <IconButton
              size="large"
              edge="start"
              color="inherit"
              aria-label={ t('aria_nav_upload') }
              sx={{ mr: 2 }}

            >
              <FileUpload />
            </IconButton>
          </Link>
          <Typography variant="h6" component="div" sx={{ flexGrow: 1 }}>
            { title }
            &nbsp;-&nbsp;
            { user.name }
          </Typography>
          <IconButton
            size="large"
            edge="start"
            color="inherit"
            aria-label={ t('aria_nav_upload') }
            sx={{ mr: 2 }}
            onClick={handleLogout}
          >
            <ExitToApp />
          </IconButton>
        </Toolbar>
      </AppBar>
    </Box>
    { children }
  </>
}
