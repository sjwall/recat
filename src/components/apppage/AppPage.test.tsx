import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';
import { store } from '../../app/store';
import { AppPage } from './AppPage';

describe('AppPage', () => {
  test('renders default', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <Provider store={store}>
          <AppPage
            title="Page title"
            user={{ name: 'users name '}}
          >
            <div>Example child</div>
          </AppPage>
        </Provider>
      </MemoryRouter>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
