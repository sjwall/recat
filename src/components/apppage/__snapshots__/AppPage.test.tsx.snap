// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppPage renders default 1`] = `
Array [
  <div
    className="MuiBox-root css-j7qwjs"
  >
    <header
      className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
    >
      <div
        className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
      >
        <a
          href="/upload"
          onClick={[Function]}
        >
          <button
            aria-label="Navigate to upload image page"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-ancrnh-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FileUploadIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z"
              />
            </svg>
          </button>
        </a>
        <div
          className="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
        >
          Page title
           - 
          users name 
        </div>
        <button
          aria-label="Navigate to upload image page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-ancrnh-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ExitToAppIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10.09 15.59 11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </button>
      </div>
    </header>
  </div>,
  <div>
    Example child
  </div>,
]
`;
